[TEAM 09] 인증을 제외한 기본 API 구현

isaac56 opened the pull request:
안녕하세요 백엔드 아이작입니다.

1주차까지 진행한 내용입니다.

## 1. DB 설계
 - Aggregate root를 크게 User, Team, Game 세 개로 나누어 설계해보았습니다.
 - Team은 Player를 갖고 있으며 여러 게임에서 재사용할 수 있도록 생각해 보았습니다.
 - Game은 진행중인 게임에 대한 정보를 갖고 있습니다.
 - Game Entity를 통하여 다음과 같은 테이블에 접근 가능합니다.
   - **batting_history**: 타자별 타율 기록)
   - **inning**: 각 이닝 별 (1회 초, 1회 말 등) 점수와 투구 기록을 갖고 있습니다.
     - Inning Entity를 통하여 pitch_history 테이블에 접근 가능합니다.
     - **pitch_history**: 어떤 투수가 어떤 타자에게 어떤 상황에서(예: 1 스트라이크 2볼인 상황) 어떤 투구(스트라이크, 볼)를 기록했는지 갖고 있습니다.
     
![](https://github.com/isaac56/baseball/blob/dev/BE/docs/database/Schema.png?raw=true)


## 2. 야구 게임의 진행을 서버에서 계산하는 방향으로 구현하였습니다.
- client는 서버에 세 가지의 결과(스트라이크, 볼, 안타) 만을 전송하고 서버는 전송받은 결과를 토대로 game 테이블을 수정합니다.
- Game entity를 통해 진행 상황에 필요한 모든 테이블에 접근 가능하기 때문에, Game entity 내부에 게임 진행에 필요한 대부분의 비즈니스 로직이 들어가있습니다.

## 3. 인증 기능을 제외한 상태에서 기본 API를 다음과 같이 구현하였습니다.
- API 문서 <https://documenter.getpostman.com/view/1562550/TzRRBTZ1>
- 게임방 목록 받기: GET /game/list
- 게임방 생성 요청: POST /game
- 게임방 참가 요청: POST /game/joining
- 게임 진행 (스트라이크, 볼, 안타) 요청: POST /game/status/pitch-result
- 게임 현황 요청(게임의 메인 뷰 업데이트): GET /game/status
- 이닝 별 스코어, 타율 기록 요청 (게임의 히스토리 뷰 업데이트): GET /game/history

## 4. AWS 배포하였습니다.
- baseball application을 위한 VPC를 생성했습니다.
- WAS용 인스턴스와 DB용 인스턴스를 나눠서 생성했습니다.
- WAS용 인스턴스는 public용 서브넷에 연결, DB는 private용 서브넷에 연결했습니다.

개발을 진행하면서 DB 설계 등 내용이 자주 변경되어 커밋 개수가 많은 점 양해부탁드립니다 ㅠㅠ
많은 지적 부탁드립니다.
항상 좋은 코드 리뷰 해주셔서 감사드립니다😁

lenaios committed:
Create .gitignore

isaac56 committed:
chore: Add gitignore contents for spring project

isaac56 committed:
[#1] Create initial database schema

요구사항을 토대로 초기 relationship model을 설계하였습니다.
만든 테이블은 다음과 같습니다.
- user
- game
- team
- player
- inning
- batting_history
- pitcher_history

isaac56 committed:
[#1] docs: Change game table schema

1. game 테이블에 1루, 2루, 3루에 해당하는 player_id 추가하였습니다.
2. user 테이블에 현재 내가 game에서 home인지 away인지 기억할 수 있도록 venue 칼럼 추가하였습니다.

isaac56 committed:
feat: Create initial spring boot project

isaac56 committed:
chore: Connect to mysql database for development

개발용 database에 연결하도록 설정하였습니다.

isaac56 committed:
[#1] docs: Change `created_time` to `index_in_inning` in pitch_history table

1. pitch_history table의 created_time을 index_in_inning으로 변경하였습니다.
2. index_in_inning은 inning 엔티티에서 pitch_history를 List로 갖기 위한 index 값입니다.

isaac56 committed:
[#2] chore: Create schema.sql

설계한 테이블을 생성하는 DDL 문 추가

isaac56 committed:
[#1] docs: Change database schema

1. player 테이블에 uniform_number 칼럼 추가
2. Game aggreagte에서 player를 가리키는 방법을 id 에서 team id + uniform number로 수정

isaac56 committed:
[#1] Change database schema

1. inning 테이블의 inning_name 칼럼명을 key_in_game으로 변경
2. batting_history 테이블에 key_in_game 칼럼 추가 (Game객체가 BattingHistory 객체를 Map으로 갖기 위한 값 추가)

isaac56 committed:
[#1] chore: Corret typo

isaac56 committed:
[#2] feat: Create Player object that is matched to player db table

isaac56 committed:
[#2] feat: Create Team object that is matched to team db table

isaac56 committed:
[#2] feat: Create Halves enum

1. 야구에서 1회"초", 1회"말"에 해당하는 Havles enum 생성
2. TOP이 초, BOTTOM이 말

isaac56 committed:
[#2] feat: Create PitchResult enum

1. 공을 던진 후의 결과에 해당하는 enum 생성
2. 안타, 스트라이크, 볼

isaac56 committed:
[#2] feat: Create PitchHistory object that is matched to pitch_history db table

isaac56 committed:
[#2] feat: Create BattingHistory object that is matched to batting_history db table

isaac56 committed:
[#2] feat: Create Inning object that is matched to inning db table

isaac56 committed:
[#2] feat: Create Game object that is matched to game db table

1. 야구 게임 한판에 해당하는 오브젝트입니다.
2. 생성자에 awayTeam, homeTeam을 인자로 받아서, 게임 시작 상황으로 초기화가 됩니다.

isaac56 committed:
[#2] feat: Create NotFoundException

데이터가 없을 경우 throw할 custom exception 생성

isaac56 committed:
[#2] feat: Create GameRepository

Aggregate root인 Game의 repository 생성

isaac56 committed:
[#2] feat: Create TeamRepository

Aggregate root인 Team의 Repository 생성

isaac56 committed:
[#2] test: Create GameRepository's create test

GameRepository 의 create 테스트 추가

isaac56 committed:
[#2] feat: Create TeamRepositoryTest for 'create test'

isaac56 committed:
[#2] chore: Add data.sql for importing initial data

isaac56 committed:
Merge pull request #3 from isaac56/feature/BE/persistant-object

[BE] 설계한 스키마에 맞게 persistent object 생성

isaac56 committed:
[#1, #2] Change database schema

1. game 테이블에 strike_count, ball_count, out_count 칼럼 추가
2. 추가한 칼럼에 맞도록 Game 객체에 프로퍼티 추가

isaac56 committed:
[#4] feat: Add getNextPlayer method that return next player

현재 선수의 등번호를 넣어주면 다음 출전 선수를 반환하는 메소드 작성

isaac56 committed:
[#4] Add method to change Game to next inning

Game 객체를 다음 이닝으로 설정해주는 메소드 추가

isaac56 committed:
[#4] refactor: Separate the method of determining defense team or attack team

isaac56 committed:
[#4] feat: Add static method that parse to key string (BattingHistory, Inning class)

1. BattingHistory와 Inning 클래스 내부에 static 메소드를 추가하였습니다.
2. 이 정적 메소드를 사용하여 외부에서도 각 클래스가 맵에 사용될 키로 변환할 수 있습니다.

isaac56 committed:
[#4] feat: Add plus method into BattingHistory

BatthingHistory의 각 카운트 값을 증가시키는 메소드 추가

isaac56 committed:
[#4] refactor: Split method that send next batter on plate from `goToNextInning`

1. 다음 타자를 타석에 세우는 과정을 sendBatterOnPlate 메소드로 분리하였습니다.
2. 다음 이닝으로 넘어가는 과정에서 말에서 초로 넘어갈 때만 1회 증가하도록 고쳤습니다.

isaac56 committed:
[#4] feat: Create method that send batter on base

타자를 출루시키는 메소드 생성

isaac56 committed:
[#4] feat: Create a method that proceed hit

타자가 안타를 쳤을 경우 게임을 진행시키는 메소드 생성하였습니다.

isaac56 committed:
[#4] feat: Create a method that proceed out

타자가 아웃되었을 경우 처리하는 메소드를 추가하였습니다.

isaac56 committed:
[#4] Create a method that proceed strike

strike의 상황을 처리하는 메소드 추가

isaac56 committed:
[#4] fix: Modify `proceedStrike` method to write pitch history

1. `proceedStrike`메소드 내부에서 투수의 스트라이크를 기록하지 않고 있었습니다.
2. 중복되는 코드를 getCurrentInning, getAttackTeamId, getDefenseTeamId 메소드로 분리했습니다.

isaac56 committed:
[#4] feat: Create a method that proceed Ball

볼일 경우 게임을 진행시키는 메소드 추가

isaac56 committed:
[#4] fix: Mdofiy `proceedOut` method to write batting history

proceedOut에서 현재 타자를 아웃처리할 때 batting history에 아웃 카운트를 증가시키지 않고 있었습니다.

isaac56 committed:
[#4] fix: Rename all methos that starts with 'get' to avoid confliciting with getter

getter와의 충돌을 피하기 위해,
get으로 시작하는 모든 메소드를 acquire로 시작하도록 변경하였습니다.

isaac56 committed:
[#4] refactor: Change `getNextPlayer`'s return type from Player to int

1. getNextPlayer 메소드가 다음 선수의 등번호를 가져올때만 사용되고 있었습니다.
2. Player를 직접 반환하지 않고 등번호를 반환하도록 수정했습니다.

isaac56 committed:
[#4] feat: Create a method that get total score

전광판의 TOP 또는 BOTTOM의 점수를 반환하는 메소드 추가

isaac56 committed:
[#4] test: Create GameTest and Add strike test method

1. Game 객체의 테스트 클래스 생성
2. 스트라이크에 따른 상황이 잘 반영되는지 테스트하는 코드 작성

isaac56 committed:
[#4] fix: Modifiy to allow all players to participate in rotation

1. 코드가 잘못되어 inning이 바뀌었을때 잘못된 등번호가 들어가고 있었습니다.
2. 모든 선수가 순환하며 출전할 수 있도록 수정하였습니다.
 - 예를 들어, away team이 수비일 때 1번이 투수였으면, 공격으로 바뀐 후부터는 2번부터 타자로 출전합니다.

isaac56 committed:
[#4] test: Add test to change inning in strike test

스트라이크 테스트 내부에서 3회 아웃으로 다음 이닝으로 넘어가는 상황까지 테스트하도록 추가

isaac56 committed:
[#4] fix: Remove setting batterUniformNumber to null

출루후에, 다음 타자는 battingUniformNumber를 참고하여 다음 순서의 타자를 내보냅니다.
출루 과정에서, battingUniformNumber를 null로 세팅하면 다음 순서를 알 수 없기 때문에 구문 삭제했습니다.

isaac56 committed:
[#4] test: Change the number of team members in the test environment to 5

테스트를 위한 팀원 수를 5:5로 변경

isaac56 committed:
[#4] test: Create ball test code

볼넷으로 득점하는 상황 테스트

isaac56 committed:
[#4] test: Create hit test method

안타로 득점하는 상황을 테스트하는 메소드 생성

isaac56 committed:
Merge pull request #5 from isaac56/feature/BE/progress-pitch

[BE] 스트라이크, 볼, 안타 세 가지 상황으로 게임을 진행시키는 로직 완성

isaac56 committed:
[#1] Create `oauth_access_token` table

사용자의 access token을 저장하고 있는 테이블 추가

isaac56 committed:
feat: Create enums

1. AWAY팀과 HOME팀을 가리키는 Venue 이넘 생성 (각각 전광판의 TOP, BOTTOM에 대응)
2. Oauth의 Resource Owner를 가리키는 ResourceOwner 이넘 생성

isaac56 committed:
[#1] chore: Set auto increment to oauth_access_token table's primary key

isaac56 committed:
[#2] feat: Create user aggregate

1. user aggregate에 속한 entity 객체들을 생성
2. UserRepository 생성
3. User의 생성 및 조회 테스트 코드 작성

isaac56 committed:
[#6] feat: Create a method that getting name from uniform number

isaac56 committed:
[#6] feat: Create common response DTO

response에 사용될 공통 형식인 ApiResult 클래스 생성

isaac56 committed:
[#6] refactor: Change some method's access modifier from private to public

1. acquireAttackTeam, acuqireDefenseTeam, acqurireCurrentInning() 메소드의 접근제어자를 public으로 변경
2. team id와 등번호로 BattingHistory를 반환해주는 메소드 추가

isaac56 committed:
[#6] feat: Add a necessary query method into UserRepository

isaac56 committed:
[#6] feat: Create PlayerDTO

isaac56 committed:
[#6] feat: Create TeamDTO

isaac56 committed:
[#6] feat: Create PitchHistoryDTO

투구 기록을 담을 DTO 생성

isaac56 committed:
[#6] feat: Create GameStatusDTO

게임 메인화면을 업데이트하기 위한 리스폰스 DTO입니다.

isaac56 committed:
[#6] feat: Create GameService

1. 야구 게임방의 서비스 레이어입니다.
2. 게임 만들기, 게임 참가하기, 게임진행하기, 게임결과받기 의 기본기능이 있습니다.

isaac56 committed:
[#6] feat: Create ApiGameController

인증과정과 Request param을 받는 걸 생략한 game관련 contoroller입니다. (추후 추가 예정)

isaac56 committed:
[#6] feat: Create GlobalExceptionHandler

isaac56 committed:
[#1] chore: Add `strike_count`, `ball_count` into `pitch_history` table

pitch_history 테이블에 투구 당시 스트라이크, 볼 상황을 기록하도록 두 개 칼럼 추가하였습니다.

isaac56 committed:
[#6] feat: Change Pitchhistory to fit modified `pitch_history` table

변경된 테이블에 맞게 PitchHistory에 strikeCount, ballCount 추가

isaac56 committed:
[#6] feat: Modify Game to write strike, ball count into pitchHistory

pitchHistory 기록할 때, 당시의 투구상황을 기록하도록 수정했습니다.

isaac56 committed:
[#6] feat: Add additional informations into PitchHistoryDTO

PitchHistoryDTO에 투구 당시의 스트라이크, 볼 카운트와 투수, 타자 정보 등을 추가로 담도록 수정

isaac56 committed:
[#6] fix: Correct a typo (BALL -> HIT)

안타를 친상황에서 PitchHistory를 BALL로 기록하고 있었습니다.

isaac56 committed:
[#6] refactor: Remove unnecessary `key_in_game` from `Inning`, `BattingHistory` entities

isaac56 committed:
[#6] refactor: Remove unnecessary `uniformNumber` from Player entity

1. Team 객체에서 uniform_number를 key로, Player를 value로 하는 Map을 들고 있습니다.
2. Player는 aggregate root인 Team을 통해서만 접근할 수 있기 때문에, Player 내부에 uniform_number는 필요 없습니다.
3. Player가 uniform_number를 갖고 있으면, spring data jdbc가 쿼리문을 생성할때 uniform_number 칼럼을 중복하여 호출합니다.

isaac56 committed:
[#6] feat: Create custom query method to get game descriptions list

게임방 목록을 보여주기 위하여 각 테이블을 조인하는 커스텀 쿼리 작성

isaac56 committed:
[#6] feat: Implements getting game list API

isaac56 committed:
[#6] refactor: Change DTO's builder access modifier from public to private

정적 팩토리 메소드 패턴을 위해서 builder의 접근제어자를 private으로 변경

lenaios committed:
Update README.md

isaac56 committed:
[#6] fix: Use `sendBatterOnPlate` method always when change batter

1. sendBatterOnPlate 메소드로 타석에 타자를 올리면 타자의 타석 수도 올려줍니다.
2. 게임 첫시작이나 다음이닝 변경 시, sendBatterOnPlate 없이 this.batterUniformNumber만 바꿔주고 있어서 타석 수가 올라가지 않고 있었습니다.

isaac56 committed:
[#6] feat: Create BattingHistoryDTO

isaac56 committed:
[#6] feat: Create TeamHistoryDTO

isaac56 committed:
[#6] feat: Create GameHistoryDTO

isaac56 committed:
[#6] Implements API that response current game history

현재 게임 기록 (스코어, 타율) 을 반환해주는 API 구현

isaac56 committed:
[#6] feat: Create requestDTO

1. 현재 만든 API 요청을 json body로 읽기위한 request DTO 생성했습니다.
2. GameController에서 json body를 읽도록 수정

lenaios committed:
Update README.md

isaac56 committed:
Merge pull request #9 from isaac56/feature/BE/implements-basic-api

[BE] 기본 API 구현 완료

isaac56 committed:
fix: Add out count missing from game history api

1song2 committed:
docs: add game play screen shots

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/request/CreateGameDTO.java:
```diff
@@ -0,0 +1,11 @@
+@Getter
+@AllArgsConstructor
+public class CreateGameDTO {
+    private int away_team_id;
+    private int home_team_id;
```
DTO라 하더라도 Java의 코딩 컨벤션을 항상 준수해 주시고요,
변수명을 카멜 케이스로 하더라도, DTO가 만들어낼 JSON의 필드명을 언더스코어 스타일로 내보낼 수 있는 설정 방식을 검색하고, 연구해 주시면 좋겠습니다.

**검색어**
- Spring Jackson Config
- Jackson annotation
- JsonProperty

isaac56 replied:
감사합니다 응답 시 DTO 변수명이 그대로 출력 되길래 큰 고민 없이 스네이크 케이스로 바꿔놨습니다.
찾아보니 카멜케이스를 json 반환시에만 스네이크케이스로 쉽게 변경할 수 있었네요!

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/response/ApiResult.java:
```diff
@@ -0,0 +1,36 @@
+@Setter
+public class ApiResult<T> {
+    private T data;
+    private String error;
```
개인적인 생각이지만 `Error` 사유가 발생할 경우를 위한 다른 클래스를 만들어서 활용하는게 더 좋을 것 같아요.
`Exception` 과 상태 코드 매칭도 될 수 있도록요.
만약 정상적인 응답이 나갔다면 이 필드는 항상 `null` 일 것 같거든요, 그리 좋은 방법으로 보이지는 않습니다.

isaac56 replied:
맞습니다 대부분 정상 응답일텐데 null을 항상 주는건 좋아보이지 않네요ㅠㅠ
지금 당장은 어떻게 바꿀수 있을지 떠오르지 않아서 천천히 고민해보겠습니다

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/response/BattingHistoryDTO.java:
```diff
@@ -0,0 +1,39 @@
+    public static BattingHistoryDTO of(Team team, BattingHistory battingHistory, int playingUniformNumber) {
+        double hitRatio = 0;
+        if (battingHistory.getAppear() != 0) {
+            hitRatio = (double) battingHistory.getHits() / (double) battingHistory.getAppear();
```
`BattingHistory` 에 위치할 로직이네요. `BattingHistory` 가 타율을 계산해주면 여기서 이렇게 로직을 전개할 필요가 없겠죠?

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/response/GameHistoryDTO.java:
```diff
@@ -0,0 +1,56 @@
+    private TeamHistoryDTO home_team;
+
+    public static GameHistoryDTO of(Game game, Team awayTeam, Team homeTeam) {
+        List<Integer> awayScores = acquireScores(game, Halves.TOP);
+        List<Integer> homeScores = acquireScores(game, Halves.BOTTOM);
```
`InningDto` 가 구성되어 있다면 아래 프라이빗 메소드에서 진행된 것처럼 복잡한 로직이 없어도 되지 않을까요?

isaac56 replied:
inning 테이블은 이닝 별로 투구 결과와 점수를 저장하기 위해 만들었습니다.

아래의 프라이빗 메소드가 다소 복잡하기는 하지만 api에 필요한 정보는 score의 리스트일 뿐이라서, 
iOS 분들과 협의한 api 형태를 바꾸면서까지 InningDto를 만들어야 할지는 모르겠습니다

애초에 설계를 잘못한 건지 score 리스트를 얻는 과정이 복잡한거 같긴 합니다 ㅠㅠ

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/response/GameHistoryDTO.java:
```diff
@@ -0,0 +1,56 @@
+        List<Integer> awayScores = acquireScores(game, Halves.TOP);
+        List<Integer> homeScores = acquireScores(game, Halves.BOTTOM);
+
+        int awayPlayingUniformNumber = game.getCurrentHalves() == Halves.TOP ?
+                game.getBatterUniformNumber() : game.getPitcherUniformNumber();
+        int homePlayingUniformNumber = game.getCurrentHalves() == Halves.BOTTOM ?
+                game.getBatterUniformNumber() : game.getPitcherUniformNumber();
```
`Game` 에서 현재 플레이중인 타자와 투수 DTO를 만들어줄 책임을 수행해야 한다고 생각합니다.

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/response/GameHistoryDTO.java:
```diff
@@ -0,0 +1,56 @@
+        List<BattingHistoryDTO> homeBattingHistories = acquireBattingHistoryDTOList(game, homeTeam, homePlayingUniformNumber);
+
+        return builder()
+                .away_team(TeamHistoryDTO.of(awayTeam.getName(), awayScores, awayBattingHistories))
+                .home_team(TeamHistoryDTO.of(homeTeam.getName(), homeScores, homeBattingHistories))
```
`.of()` 사용하신 것은 좋지만....사용해 보시니 어떤가요? 
생성자와 다른 점이 많이 느껴지시나요?
이럴때는 `Builder` 를 외부에 노출하고, 이런 곳에서도 `Builder` 사용하셔도 됩니다.

isaac56 replied:
사실 저 부분에서 생성자로 생성하는 것과 큰 차이가 없어서 고민했습니다.
다른 DTO와 통일성을 주려고 of를 사용했었는데 그냥 builder를 노출해도 되겠군요

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/response/GameHistoryDTO.java:
```diff
@@ -0,0 +1,56 @@
+
+    private static List<BattingHistoryDTO> acquireBattingHistoryDTOList(Game game, Team team, int playingUniformNumber) {
+        return game.getBattingHistoryMap().values().stream()
+                .filter(x -> x.getBatterTeamId() == team.getId())
```
`.get()` 으로 꺼내서 비교하는 것보다, `Team` 타입 파라메터를 받아 `boolean` 을 리턴하는 메소드가 구현되면 어떨까요?

```java
public boolean hasMatchingTeamId(Team team);

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/response/GameStatusDTO.java:
```diff
@@ -0,0 +1,116 @@
+                .build();
+    }
+
+    private static String acquireBatterStatus(BattingHistory batterHistory) {
+        return String.format("%d타석 %d안타", batterHistory.getAppear(), batterHistory.getHits());
+    }
```
`BattingHistory` 에 가 있어야 할 로직으로 생각되네요.

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/response/GameStatusDTO.java:
```diff
@@ -0,0 +1,116 @@
+import java.util.stream.Collectors;
+
+@Getter
+@Builder(access = AccessLevel.PRIVATE)
```
`Builder` 는 외부에 오픈되어도 됩니다.

wheejuni commented on BE/baseball/src/main/java/team9/baseball/DTO/response/PlayerDTO.java:
```diff
@@ -0,0 +1,28 @@
+    private int uniform_number;
+    private String name;
+
+    public static PlayerDTO of(Team team, Integer uniform_number) {
```
`PlayerDTO` 는 있는데 `Player` 클래스는 없는 건가요?
`Player` 클래스를 원형 그대로 파라메터로 받으면, 좀 더 좋은 구현이 가능해질텐데요. 어떻게 생각하시나요.
예를 들어 지금 애매하게 `null check` 로 진행하고 있는 유니폼 번호 유효성 체크도 `Player` 단에서 처리할 수 있을 거고요.

isaac56 replied:
맞습니다 PlayerDTO인데 Player 클래스와 연관이 없는게 어색하네요 ㅠㅠ
현재 PlayerDTO에서 필요한 정보는 team id, name, uniform number 세가지입니다.

그리고 spring data jdbc에서 Team 객체가 Player 객체들을 맵으로 갖고 있고(OneToMany), 
player 테이블의 uniform_number를 맵의 key값으로 사용합니다.
이 때, Player 객체 내부에 uniformNumber를 선언하면 DB에서 조회시, uniform_number 칼럼을 중복으로 조회하더라구요.
(Team에서 Player 맵의 key값으로 uniform number를 갖기 때문에 한번 조회하고, Player가 uniformNumber를 갖고 있기 때문에 한번 조회 하는 것 같습니다)
```
SELECT ....., **uniform_number**, **uniform_number** FROM player WHERE ...
```

그렇다 보니 Player 객체에서 uniformNumber를 갖지 않게 했고, 따라서 uniform_number는 파라미터로 받아야 할 것 같습니다.

isaac56 replied:
맞습니다 PlayerDTO인데 Player와 연관이 없다는게 이상하네요 ㅠ

말씀대로 고치는게 좋을 거 같은데 한가지 고민이 있습니다.

Team에서 Player들을 맵으로 갖고 있고(OneToMany), key값으로 player 테이블의 uniform_number를 사용하고 있습니다.
이 때, Player 객체에 uniformNumber 속성을 선언하면 uniform_number 칼럼을 중복해서 조회합니다.
(Team에서 Player 맵의 키값으로 uniform_number를 사용하기 때문에 한번, Player 객체에 uniformNumber가 선언되어 있어서 한번 조회하는것 같습니다.)
```
SELECT ....., uniform_number, uniform_number FROM player WHERE ...
```
따라서 Player 객체 내의 uniformNumber를 선언하지 않았고, 어차피 aggregate root가 Team이기 때문에 Team을 통해서만 Player에 접근할 수 있고, 따라서 uniform_number값은 Team의 playerMap의 키값으로만 접근하고 있었습니다.

PlayerDTO의 of 파라미터로 Player 하나만 받기 위해 그냥 Player 객체 내에 uniformNumber 값을 선언할지,
아니면 쿼리문에서 uniform_number를 두번 조회하지 않도록 Player 객체에 uniformNumber를 선언하지 않고 , PlayerDTO의 of 메소드에서 player와 uniformNumber를 파라미터로 받을지 고민입니다

isaac56 replied:
PlayerDTO인데 Player없이 생성하고 있었네요 😅
Player를 파라미터로 받아 생성하는게 PlayerDTO 입장에서 간결해지는 것 같습니다.

다만, 현재 API형태가 어떤 자리에 선수가 존재하지 않는다면 null을 반환하게 되어있습니다.
(1루에 주자가 없다면 Game 객체 내부의 `Integer base1UniformNumber` 값이 null이고, 마찬가지로 GameStatusDTO 내부의 `PlayerDTO base1` 값도 null을 반환해야 합니다.)

이때, Player를 파라미터로 받을 경우 
1. PlayerDTO.of 메소드에서 파라미터로 받은 Player가 null이면 null을 반환, 아니면 만든 PlayerDTO를 반환
2. PlayerDTO.of를 호출할 때, 선수번호가 null이면 null을 넣고, null이 아니면 Player 인스턴스를 넣어야 함

이렇게 두 가지 과정으로 복잡해지기 때문에 수정하진 않았습니다. (Player없이 PlayerDTO를 생성하는게 어색하긴 하지만)

wheejuni commented on BE/baseball/src/main/java/team9/baseball/service/GameService.java:
```diff
@@ -0,0 +1,124 @@
+        }
+
+        switch (pitchResult) {
+            case HIT:
+                game.proceedHit(awayTeam, homeTeam);
+                break;
+            case STRIKE:
+                game.proceedStrike(awayTeam, homeTeam);
+                break;
+            case BALL:
+                game.proceedBall(awayTeam, homeTeam);
+                break;
```
`Game.proceedBall()` 에 `pitchResult` 도 같이 넘어가면 스위치문을 없앨 수 있겠네요. 
그러면 어차피 `Game` 에 `switch` 문이 생기지 않냐고요? 그렇겠네요.
이걸 개선하려면 어떻게 해야 할까요?
`HIT, STRIKE, BALL` 에 각각 구분된 동작이 정의되어 있으면 되지 않을까요?
어떻게 정의하냐고요?
...

isaac56 replied:
HIT, STRIKE, BALL 에 따라 변하는 상황이 다르기 때문에, 코드 어딘가에서는 switch 또는 if 문으로 구분을 해야하지 않나요?
제 수준으로는 조건식 없이 저 세가지 동작을 어떻게 구분해야 할지 떠오르지 않는데 힌트좀 주실 수 있을까요?

다형성을 이용하여, `HIT, STRIKE, BALL` 각각의 구분동작을 클래스로 정의해도(같은 interface를 구현하는),
game 에서 사용할 구현체를 고를 때는 결국 조건식이 필요하기 때문에 똑같아 보입니다 ㅠ

wheejuni reviewed (changes requested):
스펙이 꽤 방대하네요. 수고 많으셨습니다. 👍 
기본적으로 아래의 원칙은 항상 준수하셔야 합니다.

- Java의 네이밍 컨벤션은 camelCase 입니다. **항상 지키셔야 합니다.**
- `switch...case` 문은 사용을 최소화합니다. _다형성_ 으로 해결할 수 있습니다. 
- 클래스가 갖고 있는 필드가 5개 이상이라면, 클래스 분리 및 책임의 위임을 고민해볼 시점입니다.

이 세 가지 원칙만 제대로 가져가셔도...수정할 코드가 꽤 많다는 것이 느껴지실 겁니다. 
제가 코드 안에 남겨드린 피드백도 참고해 주시고, 앞으로의 설계 및 코드 구현에도 이 원칙을 항상 염두에 둬 주세요. 
감사합니다. 🥇

isaac56 committed:
refactor: Change DTO's naming convention from snake case to camel case

1. DTO 객체 내부의 변수명을 스네이크 케이스에서 카멜 케이스로 변경
2. json으로 응답 시 카멜 케이스를 스네이크 케이스로 바꿔주도록 설정

isaac56 committed:
refactor: Add a getHitRatio method into BattingHistory

BattingHistory내부에 타율을 계산하여 반환하는 메소드 추가

isaac56 committed:
refactor: Add a method to get current player's number by teams into Game

isaac56 committed:
refactor: Remove unnecessary `of` method from TeamHistoryDTO

TeamHistoryDTO의 의미없는 of 메소드를 삭제하고 builder를 외부에 노출하여 생성하게 변경

isaac56 committed:
refactor: Change snake case to camel case

isaac56 committed:
refactor: Add `hasMatchingTeamId` into BattingHistory

isaac56 committed:
refactor: Add `acquireBatterStatus` method into Game

1. BattingHistory 에 getStatus 메소드 추가
2. Game에 acuqireBatterStatus 메소드 추가

isaac56 committed:
refactor: Add a `acquirePitcherStatus` method into Game

isaac56 committed:
refactor: Change PlayerDTO's `of` method to accept only Player parameter

isaac56 commented:
안녕하세요 브라이언
제가 이해할 수 있는 부분은 반영했지만
어려워서 이해하지 못한 내용은 코멘트 남겨놨습니다.

자바 코딩 컨벤션이나 클래스 책임 분리 등 기본도 못지킨 것 같아 반성하게 되었고 도움이 되었습니다.
코드 리뷰 감사드립니다😄
(실력이 부족해서 미션 완성이 빠듯할 것 같아, 큰 문제가 아니라면 approve 해주시면 감사하겠습니다ㅎㅎ)

isaac56 committed:
Revert "refactor: Change PlayerDTO's `of` method to accept only Player parameter"

This reverts commit 792dea446cc9185f45b4d5c0fcafb4ecf225b0ca.

isaac56 committed:
Merge remote-tracking branch 'origin/team9' into dev/BE
